src/renderer/object/instanced_model.rs

```diff
@@ -27,80 +28,103 @@ impl InstancedModel<ColorMaterial> {
     ///
     pub fn new(
         context: &Context,
-        transformations: &[Mat4],
+        instances: &[ModelInstance],
         cpu_mesh: &CPUMesh,
     ) -> ThreeDResult<Self> {
-        Self::new_with_material(context, transformations, cpu_mesh, ColorMaterial::default())
+        Self::new_with_material(context, instances, cpu_mesh, ColorMaterial::default())
     }
 }
 
 impl<M: ForwardMaterial> InstancedModel<M> {
     pub fn new_with_material(
         context: &Context,
-        transformations: &[Mat4],
+        instances: &[ModelInstance],
         cpu_mesh: &CPUMesh,
         material: M,
     ) -> ThreeDResult<Self> {
         let aabb = cpu_mesh.compute_aabb();
         let mut model = Self {
             context: context.clone(),
-            instance_count: 0,
             mesh: Mesh::new(context, cpu_mesh)?,
             instance_buffer1: InstanceBuffer::new(context)?,
             instance_buffer2: InstanceBuffer::new(context)?,
             instance_buffer3: InstanceBuffer::new(context)?,
+            instance_buffer4: InstanceBuffer::new(context)?,
             aabb,
             aabb_local: aabb.clone(),
             transformation: Mat4::identity(),
-            transformations: transformations.to_vec(),
+            instances: instances.to_vec(),
+            texture_transform: TextureTransform::default(),
             material,
         };
-        model.update_transformations(transformations);
-        model.update_aabb();
+        model.update_buffers();
         Ok(model)
     }
 
+    pub fn texture_transform(&mut self) -> &TextureTransform {
+        &self.texture_transform
+    }
+
+    pub fn set_texture_transform(&mut self, texture_transform: TextureTransform) {
+        self.texture_transform = texture_transform;
+    }
+
     ///
-    /// Updates the transformations applied to each model instance before they are rendered.
-    /// The model is rendered in as many instances as there are transformation matrices.
+    /// Updates instance transform and uv buffers and aabb on demand.
     ///
-    pub fn update_transformations(&mut self, transformations: &[Mat4]) {
-        self.transformations = transformations.to_vec();
-        self.instance_count = transformations.len() as u32;
+    fn update_buffers(&mut self) {
```

Comment:
This is where `!self.buffers_dirty` would shortcut. It would be set true by any function that modifies instances, then `update_buffers` can be called aggressively.